### DISCUSSION TITLE
Syntax highlighting / validation for outputs

### DISCUSSION THREAD

**Discussion by highstreeto** [NeilMacMullen/Textrude#7] [Ideas]
Currently, the output is just plain text, which is flexible, but errors / misspellings can occur when outputting for example C# source.

It would be cool, if the output field supported setting a specific language and  syntax highlighting for the generated output. Validation would be the best thing, but hard to implement for every use case (e.g. only generating fragments), so at first this could be ignored, I think.

A colleague of mine suggested using [Microsoft/monaco-editor](https://github.com/Microsoft/monaco-editor), which would include support for a lot of languages, but would be a hassle to integrate (Web-Hosting control needed).

Thoughts?

**Comment by NeilMacMullen** [comment]
Yes, I completely agree that more immediate validation would be super useful.  Since switching to AvalonEdit I'm been intending to turn on the syntax highlighting.  Unfortunately that's not great for a lot of output formats; the list of supported highlighters is relatively small and there's not deep language validation.

The link to the monaco editor is useful - I'd been looking for something like this but I think the issue with it is really that as well as the complicated hosting model you'd need to figure out a way to hook it up to the various VS code language server extensions.  Once you've done that you're quite close to having VS Code which suggests...

My first attempt at providing an interactive application was just to set up some filewatchers and to periodically write the output to a file then turn on VS code auto-save and auto-refresh and to use VS Code as the editor.  That gets you all the syntax hightlighting/validation and extension goodness you could ever hope for but turned out to be a bit clunky in practice - VS Code had a habit of throwing errors while trying to refresh the ooutput from the Textrude Engine.

So the _real_ solution would probably be to put together a VS Code extension (probably a language server) to allow VS Code to the UI host.  The Textrude Engine was written in a way that _should_ make this relatively straightforward but I don't have the time atm to dive in the the world of VS Code extensions.   I might add this to the issue list in case anyone else wants to take a run at it.

It the meantime then a compromise might be to add the ability for textrudeinteractive to write out the ouput(s) to named files whenever they change.....

**Reply by highstreeto** [reply]
Hm, a VS Code extension could be good, but also could get overwhelming (for user and developer), as (potentially) a lot of files are involved in running a template.

Auto-generating the output and viewing the generated output in another app is a bit unwieldy...

**Reply by NeilMacMullen** [reply]
Well on the one side you have the challenge of providing a fairly rich highlighting/validation mechanism (for an arbitrary set of languages!) in a TextEditor control in TextrudeeInteractive and on the other you have to come up with a way  of sensibly organising files into "projects" in  VS Code.  

I'm pretty sure the latter can be done - I use the PlatformIO extension for embedded development which does something similar though there may be better ways.  That's not to say that *also* providing a better experience in TextrudeInteractive isn't worthwhile but I'm  bit wary of getting sucked down the rabbit hole of trying to turn it into VS Code via the back door ! :-)

**Reply by highstreeto** [reply]
As a proof-of-concept I integrated [Monaco](https://github.com/Microsoft/monaco-editor) on my fork in the [`moanco`](https://github.com/highstreeto/Textrude/tree/monaco) branch. I used [WebView2](https://docs.microsoft.com/en-us/microsoft-edge/webview2/) to host *monaco* and for this to work the [WebView2 runtime](https://developer.microsoft.com/en-us/microsoft-edge/webview2/) must be installed.

*Monaco* and the starter HTML are provided to the *WebView2* by intercepting the web requests and these two files are embedded as a resource. *Monaco* itself is stored in an ZIP archive. To update the code / language in *monaco*, web messages are used, which are encoded as JSON. *Monaco* supports out-of-the-box 66 languages, so quite a range of formats are covered.

The current solution on my `monaco` branch only works for the first output and is quite hacky, so further refinements are needed.

But overall it works quite well and the integration is almost seamless:
![textrude_monaco](https://user-images.githubusercontent.com/999260/106143711-74953180-6173-11eb-87d9-cc02ac9f3695.gif)

**Reply by NeilMacMullen** [reply]
Wowo - looks fantastic - I think you have convinced me!  I'm anticipating being quite busy with "real" work the next few days - would be interested in taking this forward next week.

**Reply by NeilMacMullen** [reply]
I was intrigued enough by this that just pushed some changes to better componentise the output section.  There is now an OutputPane user control which has a built in format selection (which is persisted via the project mechanism).  If you replace the avaloneditor  in this control with your Monaco stuff, you should get customisable formats for all output panes.

**Reply by NeilMacMullen** [reply]
Parking this here so I don't lose it....   https://stackoverflow.com/questions/64740327/detect-if-webview2-is-installed-on-clients-machine-vb-net   Would be nice to warn users if the runtime is missing.

**Reply by highstreeto** [reply]
Ok, I will look into integrating *monaco* into the new `OutputPanes`.
One problem might be, that the *WebView2* is quite heavyweight (uses a separate process) and spawning one per tab could get problematic. I have to test this, to see if there is a problem.
Another solution would be to handle all outputs in one *monaco* instance, as *monaco* supports holding onto different text models (I will have to look into this, to see if it is feasible).

**Reply by NeilMacMullen** [reply]
Nice. Yes - I had the same thought about sharing a single instance across "pseudo tabs" as well. That could be a future optimisation though.

**Reply by NeilMacMullen** [reply]
>That could be a future optimisation though.

Also worth noting the number of model(input)  and ouput panes is now controlled by the user on a per-project basis so in practice I'd expect most projects at this stage to have just one model and one output.  Hence the overhead of a  webview per tab is possibly not so relevant as when we always created 3 models and 3 outputs

**Reply by highstreeto** [reply]
Yeah, same thought here.

So, I integrated *monaco* into the new `OutputPanes`, which is now on my [`monaco`](https://github.com/highstreeto/Textrude/tree/monaco) branch. It works rather well and was quite easy to integrate :)

![textrude_monaco_v2](https://user-images.githubusercontent.com/999260/106301029-79cbac80-6257-11eb-8aab-697862e0cf42.gif)

**Reply by NeilMacMullen** [reply]
Nice!  I know I'm getting greedy now but could it also work in the InputPanes?  (I know you'd need to manage keyboard input which is a new feature)  It would be nice to get syntax highlighting on the model! Regardless. very much looking forward to merging this ! :-)

**Reply by highstreeto** [reply]
It is possible... ;)

![textrude_monaco_v3](https://user-images.githubusercontent.com/999260/106310861-81458280-6264-11eb-9424-82c53f294da3.gif)

**Reply by NeilMacMullen** [reply]
Bravo!

**Reply by NeilMacMullen** [reply]
BTW -if you haven't already worked it out, fix for unrendered tab is to add this line in TabControlManager.AddPane...

 ```
  _tab.SelectedIndex = _tab.Items.Count - 1; //ensure tab is rendered
            return pane;
```

I have it queued in a commit here and will push later.

**Reply by NeilMacMullen** [reply]
FYI - I have just pushed some changes to better support file-backed model/output panes.  Hopefully won't conflict with your pending changes.
